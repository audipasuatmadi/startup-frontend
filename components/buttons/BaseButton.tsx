import React from 'react';
import Link from 'next/link';

export interface BaseButtonProps {
  disabled?: boolean;
  href?: string;
  children?: any;

  className?: string;
  type?: 'button' | 'submit' | 'reset';

  onClick?: () => void;
}

/**
 *
 * @version 0.1.0
 * @description this component is the base of all buttons. It does basic things such as deciding what to do if disabled, return Link or button, and many more!
 * @author Putu Audi Pasuatmadi
 */
const BaseButton = ({
  href,
  disabled,
  onClick,
  children,
  type,
  className,
}: BaseButtonProps) => {
  const finalStyles = `${disabled ? 'cursor-default' : ''}${
    className ? ` ${className}` : ''
  }`;

  return href && !disabled ? (
    <Link href={href}>
      <a className={finalStyles !== '' ? finalStyles : undefined}>{children}</a>
    </Link>
  ) : (
    <button
      className={finalStyles !== '' ? finalStyles : undefined}
      disabled={disabled}
      type={type}
      onClick={!disabled && onClick}>
      {children}
    </button>
  );
};

export default BaseButton;
