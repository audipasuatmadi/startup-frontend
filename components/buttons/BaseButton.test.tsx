import React from 'react'
import { shallow } from 'enzyme'
import BaseButton from './BaseButton'

describe('BaseButton functionality tests', () => {
  it('should clickable if not disabled', () => {
    const mockFunction = jest.fn()
    const wrapper = shallow(<BaseButton onClick={mockFunction} />)
    
    wrapper.find('button').simulate('click')
    expect(mockFunction).toHaveBeenCalled()
    wrapper.find('button').simulate('click')
    expect(mockFunction).toHaveBeenCalledTimes(2)
  });

  it('should be not clickable if disabled', () => {
    const mockFunction = jest.fn()
    const wrapper = shallow(<BaseButton disabled onClick={mockFunction} />)
    
    wrapper.find('button').simulate('click')
    expect(mockFunction).not.toHaveBeenCalled()
  })
})

describe('BaseButton rendering tests', () => {
  it('should render properly as a button without args', () => {
    const wrapper = shallow(<BaseButton />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should render properly as a button when disabled, even with a href', () => {
    const wrapper = shallow(<BaseButton disabled href="/home" />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should render properly as a link, with href args without disabled', () => {
    const wrapper = shallow(<BaseButton href="/home" />)
    expect(wrapper).toMatchSnapshot()
  })

})