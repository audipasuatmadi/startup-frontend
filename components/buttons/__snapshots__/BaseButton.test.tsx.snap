// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseButton rendering tests should render properly as a button when disabled, even with a href 1`] = `
<button
  className="cursor-default"
  disabled={true}
  onClick={false}
  onMouseDown={false}
/>
`;

exports[`BaseButton rendering tests should render properly as a button without args 1`] = `<button />`;

exports[`BaseButton rendering tests should render properly as a link, with href args without disabled 1`] = `
<Link
  href="/home"
>
  <a />
</Link>
`;
